src/main/java/io/debezium/connector/vitess/VitessConnectorTask.java

```diff
@@ -148,52 +150,104 @@ protected ChangeEventSourceCoordinator<VitessPartition, VitessOffsetContext> sta
     }
 
     @VisibleForTesting
-    protected Configuration getConfigWithOffsets(Configuration config) {
+    public Configuration getConfigWithOffsets(Configuration config) {
         VitessConnectorConfig connectorConfig = new VitessConnectorConfig(config);
         if (connectorConfig.offsetStoragePerTask()) {
-            int gen = connectorConfig.getOffsetStorageTaskKeyGen();
-            int prevGen = gen - 1;
-            Map<String, String> prevGtidsPerShard = VitessConnector.getGtidPerShardFromStorage(
-                    context.offsetStorageReader(),
-                    connectorConfig,
-                    connectorConfig.getPrevNumTasks(),
-                    prevGen,
-                    true);
-            LOGGER.info("prevGtidsPerShard {}", prevGtidsPerShard);
-            Map<String, String> gtidsPerShard = VitessConnector.getGtidPerShardFromStorage(
-                    context.offsetStorageReader(),
-                    connectorConfig,
-                    connectorConfig.getVitessTotalTasksConfig(),
-                    gen,
-                    false);
-            LOGGER.info("gtidsPerShard {}", gtidsPerShard);
-            List<String> shards = connectorConfig.getVitessTaskKeyShards();
-            Map<String, String> configGtidsPerShard = getConfigGtidsPerShard(connectorConfig, shards);
-            LOGGER.info("configGtidsPerShard {}", configGtidsPerShard);
-            final String keyspace = connectorConfig.getKeyspace();
-
-            List<Vgtid.ShardGtid> shardGtids = new ArrayList<>();
-            for (String shard : shards) {
-                String gtidStr;
-                if (gtidsPerShard.containsKey(shard)) {
-                    gtidStr = gtidsPerShard.get(shard);
-                    LOGGER.info("Using offsets from current gen: shard {}, gen {}, gtid {}", shard, gen, gtidStr);
-                }
-                else if (prevGtidsPerShard.containsKey(shard)) {
-                    gtidStr = prevGtidsPerShard.get(shard);
-                    LOGGER.warn("Using offsets from previous gen: shard {}, gen {}, gtid {}", shard, gen, gtidStr);
-                }
-                else {
-                    gtidStr = configGtidsPerShard.getOrDefault(shard, null);
-                    LOGGER.warn("Using offsets from config: shard {}, gtid {}", shard, gtidStr);
-                }
-                shardGtids.add(new Vgtid.ShardGtid(keyspace, shard, gtidStr));
+            config = getVitessTaskValuePerShard(config, connectorConfig, VitessOffsetRetriever.ValueType.GTID);
+            if (VitessOffsetRetriever.isShardEpochMapEnabled(connectorConfig)) {
```

Comment:
This looks like should be if ... else structure otherwise line 156 was executed in vain.

Reply:
The `getVitessTaskValuePerShard` actually modifies the map (since it passes it in as an arg). However, that is not clear (not well named) and better to not have hidden side effects on the config. I updated so it simply returns the value and then the config editing is done in this function.